src/graph/scheduler/AsyncMsgNotifyBasedScheduler.cpp

```diff
@@ -225,17 +239,94 @@ void AsyncMsgNotifyBasedScheduler::notifyError(std::vector<folly::Promise<Status
+
+  return std::move(exeStatus).thenValue([this, executor](Status s) {
+    if (!s.ok()) {
+      DLOG(INFO) << formatPrettyId(executor) << " failed with: " << s.toString();
+      setFailStatus(s);
+      removeExecuting(executor);
+      return Status::from(s);
+    }
+    auto ret = executor->close();
+    removeExecuting(executor);
+    return ret;
   });
 }
 
+folly::Future<Status> AsyncMsgNotifyBasedScheduler::runExecute(Executor* executor) const {
+  // catch Executor::execute here, upward call stack should only get Status, no exceptions.
+  try {
+    addExecuting(executor);
+    if (hasFailStatus()) return failedStatus_.value();
+    folly::Future<Status> status = Status::OK();
+    {
+      memory::MemoryCheckGuard guard;
+      status = executor->execute();
+    }
+    return std::move(status).thenError(folly::tag_t<std::bad_alloc>{}, [](const std::bad_alloc&) {
+      return folly::makeFuture<Status>(Status::GraphMemoryExceeded(
+          "(%d)", static_cast<int32_t>(nebula::cpp2::ErrorCode::E_GRAPH_MEMORY_EXCEEDED)));
+    });
+  } catch (std::bad_alloc& e) {
+    return folly::makeFuture<Status>(Status::GraphMemoryExceeded(
+        "(%d)", static_cast<int32_t>(nebula::cpp2::ErrorCode::E_GRAPH_MEMORY_EXCEEDED)));
+  } catch (std::exception& e) {
+    return folly::makeFuture<Status>(Status::Error("%s", e.what()));
+  } catch (...) {
+    return folly::makeFuture<Status>(Status::Error("unknown error"));
+  }
+}
+
+void AsyncMsgNotifyBasedScheduler::addExecuting(Executor* executor) const {
+  std::unique_lock<std::mutex> lck(emtx_);
+  executing_++;
+  DLOG(INFO) << formatPrettyId(executor) << " add " << executing_;
+}
+
+void AsyncMsgNotifyBasedScheduler::removeExecuting(Executor* executor) const {
+  std::unique_lock<std::mutex> lck(emtx_);
+  executing_--;
+  DLOG(INFO) << formatPrettyId(executor) << "remove: " << executing_;
+  cv_.notify_one();
+}
+
+void AsyncMsgNotifyBasedScheduler::setFailStatus(Status status) const {
+  std::unique_lock<std::mutex> lck(smtx_);
+  if (!failedStatus_.has_value()) {
+    failedStatus_ = status;
+  }
+}
+
+bool AsyncMsgNotifyBasedScheduler::hasFailStatus() const {
+  std::unique_lock<std::mutex> lck(smtx_);
+  return failedStatus_.has_value();
+}
+
+std::string AsyncMsgNotifyBasedScheduler::formatPrettyId(Executor* executor) {
+  return fmt::format("[{},{}]", executor->name(), executor->id());
+}
+
+std::string AsyncMsgNotifyBasedScheduler::formatPrettyDependencyTree(Executor* root) {
+  std::stringstream ss;
+  size_t spaces = 0;
+  appendExecutor(spaces, root, ss);
+  return ss.str();
+}
+
+void AsyncMsgNotifyBasedScheduler::appendExecutor(size_t spaces,
+                                                  Executor* executor,
+                                                  std::stringstream& ss) {
+  ss << std::string(spaces, ' ') << formatPrettyId(executor) << std::endl;
+  for (auto depend : executor->depends()) {
+    appendExecutor(spaces + 1, depend, ss);
+  }
```

Comment:
do the loop body executors need to be considered here?

Reply:
> do the loop body executors need to be considered here?

no, just print a original dependency tree.